client/src/main/java/io/getstream/chat/android/client/utils/observable/ChatEventsObservable.kt

```diff
@@ -0,0 +1,102 @@
+package io.getstream.chat.android.client.utils.observable
+
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.ConnectedEvent
+import io.getstream.chat.android.client.events.ConnectingEvent
+import io.getstream.chat.android.client.events.DisconnectedEvent
+import io.getstream.chat.android.client.events.ErrorEvent
+import io.getstream.chat.android.client.models.EventType
+import io.getstream.chat.android.client.socket.ChatSocket
+import io.getstream.chat.android.client.socket.ChatSocketService
+import io.getstream.chat.android.client.socket.SocketListener
+import java.util.Date
+
+internal class ChatEventsObservable(private val socket: ChatSocket) {
+
+    private val subscriptions = mutableSetOf<EventSubscription>()
+    private var eventsMapper = EventsMapper(this)
+
+    private fun onNext(event: ChatEvent) {
+        subscriptions.forEach { subscription ->
+            if (!subscription.isDisposed) {
+                subscription.onNext(event)
+            }
+        }
+        subscriptions.removeAll(Disposable::isDisposed)
+        checkIfEmpty()
+    }
+
+    private fun checkIfEmpty() {
+        if (subscriptions.isEmpty()) {
+            socket.removeListener(eventsMapper)
+        }
+    }
+
+    fun subscribe(
+        filter: (ChatEvent) -> Boolean = { true },
+        listener: (ChatEvent) -> Unit
+    ): Disposable {
+        return addSubscription(SubscriptionImpl(filter, listener))
+    }
+
+    fun subscribeSingle(
+        filter: (ChatEvent) -> Boolean = { true },
+        listener: (ChatEvent) -> Unit
+    ): Disposable {
+        return addSubscription(SingleSubscriptionImpl(filter, listener))
+    }
+
+    private fun addSubscription(subscription: EventSubscription): Disposable {
+        if (subscriptions.isEmpty()) {
+            // add listener to socket events only once
+            socket.addListener(eventsMapper)
+        }
+
+        subscriptions.add(subscription)
```

Comment:
Is it thread safe? This could be invoked from any thread. I think it should be syncrnozied

Reply:
You're correct, this is not thread safe. For now, it's a reimplementation of the old API, which also did no thread handling at all. It's definitely something we should look into later on though.